
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>ArchR I/O &#8212; chame  documentation</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Ranges" href="ranges.html" />
    <link rel="prev" title="10x Genomics I/O" href="10x_io.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/chame_logo.svg" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="index.html">
   Examples
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="10x_io.html">
     10x Genomics I/O
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     ArchR I/O
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="ranges.html">
     Ranges
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="chromvar.html">
     chromVAR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="seq.html">
     Utility functions for sequences
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../api.html">
   API
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../references.html">
   References
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Theme by the <a href="https://ebp.jupyterbook.org">Executable Book Project</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/examples/archr_io.ipynb.txt"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#ArchR:-Creating-Arrow-Files">
   ArchR: Creating Arrow Files
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#Reading-matrices-from-Arrow-files">
   Reading matrices from Arrow files
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Tiles">
     Tiles
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Gene-scores">
     Gene scores
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#All-matrices">
     All matrices
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>ArchR I/O</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#ArchR:-Creating-Arrow-Files">
   ArchR: Creating Arrow Files
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#Reading-matrices-from-Arrow-files">
   Reading matrices from Arrow files
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Tiles">
     Tiles
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Gene-scores">
     Gene scores
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#All-matrices">
     All matrices
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars and line breaks on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
    white-space: pre;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="ArchR-I/O">
<h1>ArchR I/O<a class="headerlink" href="#ArchR-I/O" title="Permalink to this headline">#</a></h1>
<p><a class="reference external" href="https://www.archrproject.com/">ArchR</a> is an R package for scATAC-seq data analysis <a class="reference external" href="https://www.nature.com/articles/s41588-021-00790-6">(Granja et al., 2021)</a>.</p>
<p><code class="docutils literal notranslate"><span class="pre">chame</span></code> comes with a reader for ArchR’s <a class="reference external" href="https://www.archrproject.com/bookdown/what-is-an-arrow-file-archrproject.html">Arrow files</a>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from chame.io import read_arrow
</pre></div>
</div>
</div>
<section id="ArchR:-Creating-Arrow-Files">
<h2>ArchR: Creating Arrow Files<a class="headerlink" href="#ArchR:-Creating-Arrow-Files" title="Permalink to this headline">#</a></h2>
<p>For resproducibility, we will follow ArchR’s own <a class="reference external" href="https://www.archrproject.com/bookdown/creating-arrow-files.html">tutorial</a> on creating Arrow files.</p>
<p>Here, we use <code class="docutils literal notranslate"><span class="pre">rpy2</span></code> for R calls, and it is assumed that the corresponding R environemnt is configured and has ArchR installed.</p>
<p>In practice, one would most likely use existing Arrow files. So feel free to skip this section!</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import os
os.chdir(&quot;../../&quot;)
if not os.path.exists(&quot;archr&quot;):
    os.mkdir(&quot;archr&quot;)
os.chdir(&quot;archr/&quot;)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from rpy2.robjects.packages import importr

archr = importr(&quot;ArchR&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
R[write to console]:
                                                   / |
                                                 /    \
            .                                  /      |.
            \\\                              /        |.
              \\\                          /           `|.
                \\\                      /              |.
                  \                    /                |\
                  \\#####\           /                  ||
                ==###########&gt;      /                   ||
                 \\##==......\    /                     ||
            ______ =       =|__ /__                     ||      \\\
        ,--&#39; ,----`-,__ ___/&#39;  --,-`-===================##========&gt;
       \               &#39;        ##_______ _____ ,--,__,=##,__   ///
        ,    __==    ___,-,__,--&#39;#&#39;  ===&#39;      `-&#39;    | ##,-/
        -,____,---&#39;       \\####\\________________,--\\_##,/
           ___      .______        ______  __    __  .______
          /   \     |   _  \      /      ||  |  |  | |   _  \
         /  ^  \    |  |_)  |    |  ,----&#39;|  |__|  | |  |_)  |
        /  /_\  \   |      /     |  |     |   __   | |      /
       /  _____  \  |  |\  \\___ |  `----.|  |  |  | |  |\  \\___.
      /__/     \__\ | _| `._____| \______||__|  |__| | _| `._____|


R[write to console]: Setting default number of Parallel threads to 4.

R[write to console]: Setting addArchRVerbose = FALSE

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>archr.addArchRGenome(&quot;hg19&quot;);
archr.addArchRThreads(threads = 6);
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
R[write to console]: Setting default genome to Hg19.

R[write to console]: Setting default number of Parallel threads to 6.

</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>inputFiles = archr.getTutorialData(&quot;Hematopoiesis&quot;)

arrowFiles = archr.createArrowFiles(
  inputFiles=inputFiles,
  sampleNames=inputFiles.names,
  minTSS=4,
  filterFrags=1000,
  addTileMat=True,
  addGeneScoreMat=True,
)
</pre></div>
</div>
</div>
</section>
<section id="Reading-matrices-from-Arrow-files">
<h2>Reading matrices from Arrow files<a class="headerlink" href="#Reading-matrices-from-Arrow-files" title="Permalink to this headline">#</a></h2>
<section id="Tiles">
<h3>Tiles<a class="headerlink" href="#Tiles" title="Permalink to this headline">#</a></h3>
<p>With a matrix specified, an AnnData object will be returned. For <em>tiles</em>, it would have dimensions <em>n_cells</em> by <em>n_tiles</em>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>tiles = read_arrow(&quot;scATAC_PBMC_R1.arrow&quot;, matrix=&quot;tiles&quot;)
tiles
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2453 × 6072620
    obs: &#39;BlacklistRatio&#39;, &#39;CellNames&#39;, &#39;Completed&#39;, &#39;Date&#39;, &#39;NucleosomeRatio&#39;, &#39;PassQC&#39;, &#39;PromoterRatio&#39;, &#39;ReadsInBlacklist&#39;, &#39;ReadsInPromoter&#39;, &#39;ReadsInTSS&#39;, &#39;Sample&#39;, &#39;TSSEnrichment&#39;, &#39;nDiFrags&#39;, &#39;nFrags&#39;, &#39;nMonoFrags&#39;, &#39;nMultiFrags&#39;
    var: &#39;Chromosome&#39;, &#39;idx&#39;, &#39;Start&#39;, &#39;End&#39;
    uns: &#39;params&#39;
</pre></div></div>
</div>
<p>As displayed, this AnnData object has cell metadata included together with the information about features (tiles):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>tiles.obs.head(2)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>BlacklistRatio</th>
      <th>CellNames</th>
      <th>Completed</th>
      <th>Date</th>
      <th>NucleosomeRatio</th>
      <th>PassQC</th>
      <th>PromoterRatio</th>
      <th>ReadsInBlacklist</th>
      <th>ReadsInPromoter</th>
      <th>ReadsInTSS</th>
      <th>Sample</th>
      <th>TSSEnrichment</th>
      <th>nDiFrags</th>
      <th>nFrags</th>
      <th>nMonoFrags</th>
      <th>nMultiFrags</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CCGTGAGAGACCATAA-1</th>
      <td>0.010204</td>
      <td>CCGTGAGAGACCATAA-1</td>
      <td>Finished</td>
      <td>2022-06-10</td>
      <td>3.578472</td>
      <td>1.0</td>
      <td>0.064401</td>
      <td>1861.0</td>
      <td>11746.0</td>
      <td>2926.0</td>
      <td>scATAC_PBMC_R1</td>
      <td>6.652</td>
      <td>47936.0</td>
      <td>91194.0</td>
      <td>19918.0</td>
      <td>23340.0</td>
    </tr>
    <tr>
      <th>TGCACCTTCGGCTATA-1</th>
      <td>0.008097</td>
      <td>TGCACCTTCGGCTATA-1</td>
      <td>Finished</td>
      <td>2022-06-10</td>
      <td>4.809345</td>
      <td>1.0</td>
      <td>0.042707</td>
      <td>1351.0</td>
      <td>7126.0</td>
      <td>1571.0</td>
      <td>scATAC_PBMC_R1</td>
      <td>4.699</td>
      <td>46671.0</td>
      <td>83428.0</td>
      <td>14361.0</td>
      <td>22396.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>tiles.var.head()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Chromosome</th>
      <th>idx</th>
      <th>Start</th>
      <th>End</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>chr1:0-500</th>
      <td>chr1</td>
      <td>1</td>
      <td>0</td>
      <td>500</td>
    </tr>
    <tr>
      <th>chr1:500-1000</th>
      <td>chr1</td>
      <td>2</td>
      <td>500</td>
      <td>1000</td>
    </tr>
    <tr>
      <th>chr1:1000-1500</th>
      <td>chr1</td>
      <td>3</td>
      <td>1000</td>
      <td>1500</td>
    </tr>
    <tr>
      <th>chr1:1500-2000</th>
      <td>chr1</td>
      <td>4</td>
      <td>1500</td>
      <td>2000</td>
    </tr>
    <tr>
      <th>chr1:2000-2500</th>
      <td>chr1</td>
      <td>5</td>
      <td>2000</td>
      <td>2500</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>↑ As AnnData currently requires string indices for both cells and features, tiles here are automatically named by <code class="docutils literal notranslate"><span class="pre">chame</span></code> according to their coordinates.</p>
<p>Notably, <code class="docutils literal notranslate"><span class="pre">chame</span></code> strives to return data frames that can be readily used to create PyRanges objects:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pyranges import PyRanges
pr = PyRanges(tiles.var)

pr[&quot;chr1&quot;, 0:4200]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
+--------------+-----------+-----------+-----------+
| Chromosome   | idx       | Start     | End       |
| (category)   | (int32)   | (int32)   | (int32)   |
|--------------+-----------+-----------+-----------|
| chr1         | 1         | 0         | 500       |
| chr1         | 2         | 500       | 1000      |
| chr1         | 3         | 1000      | 1500      |
| chr1         | 4         | 1500      | 2000      |
| ...          | ...       | ...       | ...       |
| chr1         | 6         | 2500      | 3000      |
| chr1         | 7         | 3000      | 3500      |
| chr1         | 8         | 3500      | 4000      |
| chr1         | 9         | 4000      | 4500      |
+--------------+-----------+-----------+-----------+
Unstranded PyRanges object has 9 rows and 4 columns from 1 chromosomes.
For printing, the PyRanges was sorted on Chromosome.
</pre></div></div>
</div>
<p>Params for the <em>TileMatrix</em> from the Arrow file are carefully preserved and put into <code class="docutils literal notranslate"><span class="pre">.uns</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>tiles.uns[&quot;params&quot;][:3]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[{&#39;seqnames&#39;: &#39;chr1&#39;, &#39;length&#39;: 249250621, &#39;tileSize&#39;: 500.0},
 {&#39;seqnames&#39;: &#39;chr2&#39;, &#39;length&#39;: 243199373, &#39;tileSize&#39;: 500.0},
 {&#39;seqnames&#39;: &#39;chr3&#39;, &#39;length&#39;: 198022430, &#39;tileSize&#39;: 500.0}]
</pre></div></div>
</div>
</section>
<section id="Gene-scores">
<h3>Gene scores<a class="headerlink" href="#Gene-scores" title="Permalink to this headline">#</a></h3>
<p>For <em>gene scores</em>, variables will be respectively genes:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>gene_scores = read_arrow(&quot;scATAC_PBMC_R1.arrow&quot;, matrix=&quot;gene_scores&quot;)
gene_scores
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2453 × 23127
    obs: &#39;BlacklistRatio&#39;, &#39;CellNames&#39;, &#39;Completed&#39;, &#39;Date&#39;, &#39;NucleosomeRatio&#39;, &#39;PassQC&#39;, &#39;PromoterRatio&#39;, &#39;ReadsInBlacklist&#39;, &#39;ReadsInPromoter&#39;, &#39;ReadsInTSS&#39;, &#39;Sample&#39;, &#39;TSSEnrichment&#39;, &#39;nDiFrags&#39;, &#39;nFrags&#39;, &#39;nMonoFrags&#39;, &#39;nMultiFrags&#39;
    var: &#39;seqnames&#39;, &#39;start&#39;, &#39;end&#39;, &#39;strand&#39;, &#39;name&#39;, &#39;idx&#39;
    uns: &#39;params&#39;
</pre></div></div>
</div>
<p>Params for the <em>GeneScoresMatrix</em> from the Arrow file are also put into <code class="docutils literal notranslate"><span class="pre">.uns</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>gene_scores.uns[&quot;params&quot;][:3]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[{&#39;extendUpstream&#39;: 1000.0,
  &#39;extendDownstream&#39;: 1000.0,
  &#39;geneUpstream&#39;: 1000.0,
  &#39;geneDownstream&#39;: 1000.0,
  &#39;scaleTo&#39;: 10000.0,
  &#39;tileSize&#39;: 500.0,
  &#39;ceiling&#39;: 4.0,
  &#39;geneModel&#39;: b&#39;exp(-abs(x)/5000) + exp(-1)&#39;},
 {&#39;extendUpstream&#39;: 100000.0,
  &#39;extendDownstream&#39;: 100000.0,
  &#39;geneUpstream&#39;: 100000.0,
  &#39;geneDownstream&#39;: 100000.0,
  &#39;scaleTo&#39;: 10000.0,
  &#39;tileSize&#39;: 500.0,
  &#39;ceiling&#39;: 4.0,
  &#39;geneModel&#39;: b&#39;exp(-abs(x)/5000) + exp(-1)&#39;}]
</pre></div></div>
</div>
</section>
<section id="All-matrices">
<h3>All matrices<a class="headerlink" href="#All-matrices" title="Permalink to this headline">#</a></h3>
<p>If no matrix is specified, all available matrices are read into a MuData object by default:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mdata = read_arrow(&quot;scATAC_PBMC_R1.arrow&quot;)
mdata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<pre>MuData object with n_obs × n_vars = 2453 × 6095747
  obs:      &#x27;BlacklistRatio&#x27;, &#x27;CellNames&#x27;, &#x27;Completed&#x27;, &#x27;Date&#x27;, &#x27;NucleosomeRatio&#x27;, &#x27;PassQC&#x27;, &#x27;PromoterRatio&#x27;, &#x27;ReadsInBlacklist&#x27;, &#x27;ReadsInPromoter&#x27;, &#x27;ReadsInTSS&#x27;, &#x27;Sample&#x27;, &#x27;TSSEnrichment&#x27;, &#x27;nDiFrags&#x27;, &#x27;nFrags&#x27;, &#x27;nMonoFrags&#x27;, &#x27;nMultiFrags&#x27;
  var:      &#x27;idx&#x27;
  2 modalities
    tiles:  2453 x 6072620
      var:  &#x27;Chromosome&#x27;, &#x27;idx&#x27;, &#x27;Start&#x27;, &#x27;End&#x27;
      uns:  &#x27;params&#x27;
    gene_scores:    2453 x 23127
      var:  &#x27;seqnames&#x27;, &#x27;start&#x27;, &#x27;end&#x27;, &#x27;strand&#x27;, &#x27;name&#x27;, &#x27;idx&#x27;
      uns:  &#x27;params&#x27;</pre></div>
</div>
</section>
</section>
</section>


              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="10x_io.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">10x Genomics I/O</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="ranges.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Ranges</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Danila Bredikhin<br/>
  
      &copy; Copyright 2022, Danila Bredikhin..<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>